python/ray/ml/preprocessors/encoder.py

```diff
@@ -0,0 +1,129 @@
+from typing import List, Dict
+
+import pandas as pd
+
+from ray.data import Dataset
+from ray.data.aggregate import UniqueValues
+from ray.ml.preprocessor import Preprocessor
+
+
+class OrdinalEncoder(Preprocessor):
+    """Encode values within columns as ordered integer values.
+
+    Currently, order within a column is based on the values from the fitted
+    dataset in sorted order.
+
+    Transforming values not included in the fitted dataset will be encoded as ``None``.
+
+    Args:
+        columns: The columns that will individually be encoded.
+    """
+
+    def __init__(self, columns: List[str]):
+        # TODO: allow user to specify order of values within each column.
+        super().__init__()
+        self.columns = columns
+
+    def _fit(self, dataset: Dataset) -> Preprocessor:
+        aggregates = [UniqueValues(col) for col in self.columns]
+        stats = dataset.aggregate(*aggregates)
+        self.stats = {row: _sorted_value_indices(stats[row]) for row in stats}
+        return self
+
+    def _transform_pandas(self, df: pd.DataFrame):
+        def column_ordinal_encoder(s: pd.Series):
+            s_values = self.stats[f"unique_values({s.name})"]
+            return s.map(s_values)
+
+        df.loc[:, self.columns] = df.loc[:, self.columns].transform(
+            column_ordinal_encoder
+        )
+        return df
+
+    def __repr__(self):
+        return f"<Encoder columns={self.columns} stats={self.stats}>"
+
+
+class OneHotEncoder(Preprocessor):
+    """Encode columns as new columns using one-hot encoding.
+
+    The transformed dataset will have a new column in the form ``{column}_{value}``
+    for each of the values from the fitted dataset. The value of a column will
+    be set to 1 if the value matches, otherwise 0.
+
+    Transforming values not included in the fitted dataset will result in all
+    of the encoded column values to be 0.
+
+    Args:
+        columns: The columns that will individually be encoded.
+    """
+
+    def __init__(self, columns: List[str]):
+        super().__init__()
+        self.columns = columns
+
+    def _fit(self, dataset: Dataset) -> Preprocessor:
+        aggregates = [UniqueValues(col) for col in self.columns]
+        stats = dataset.aggregate(*aggregates)
+        self.stats = {row: _sorted_value_indices(stats[row]) for row in stats}
+        return self
+
+    def _transform_pandas(self, df: pd.DataFrame):
+        for column in self.columns:
+            column_values = self.stats[f"unique_values({column})"]
+            for column_value in column_values:
+                df[f"{column}_{column_value}"] = df[column].map(
+                    lambda x: int(x == column_value)
+                )
```

Comment:
```suggestion
            for column_value in column_values:
                df[f"{column}_{column_value}"] = (df[column] == column_value).astype(int)
```

Faster than `map`

Reply:
Nit: I'm a fan of the map over default dict route for one-hot encoding, not sure how the speed compares.
```suggestion
            for column_value in column_values:
                one_hot = collections.default_dict(int, **{column_value: 1})
                df[f"{column}_{column_value}"] = df[column].map(one_hot)
```
All of these approaches look good to me, no strong opinion. so if @Yard1's is fastest, that sounds good to me.

Reply:
Going with Antoni's for the time being assuming its faster and functionally equivalent.